polkadot/xcm/src/v5/junction.rs

```diff
@@ -0,0 +1,322 @@
+	// the bound for `Copy`.
+	GeneralKey { length: u8, data: [u8; 32] },
+	/// The unambiguous child.
+	///
+	/// Not currently used except as a fallback when deriving context.
+	OnlyChild,
+	/// A pluralistic body existing within consensus.
+	///
+	/// Typical to be used to represent a governance origin of a chain, but could in principle be
+	/// used to represent things such as multisigs also.
+	Plurality { id: BodyId, part: BodyPart },
+	/// A global network capable of externalizing its own consensus. This is not generally
+	/// meaningful outside of the universal level.
+	GlobalConsensus(NetworkId),
+}
+
+/// The genesis hash of the Westend testnet. Used to identify it.
+pub const WESTEND_GENESIS_HASH: [u8; 32] =
+	hex!["e143f23803ac50e8f6f8e62695d1ce9e4e1d68aa36c1cd2cfd15340213f3423e"];
+
+/// The genesis hash of the Rococo testnet. Used to identify it.
+pub const ROCOCO_GENESIS_HASH: [u8; 32] =
+	hex!["6408de7737c59c238890533af25896a2c20608d8b380bb01029acb392781063e"];
+
+// TODO: What to use here?
+pub const WOCOCO_GENESIS_HASH: [u8; 32] =
+	hex!["0000000000000000000000000000000000000000000000000000000000000000"];
+
+/// A global identifier of a data structure existing within consensus.
+///
+/// Maintenance note: Networks with global consensus and which are practically bridgeable within the
+/// Polkadot ecosystem are given preference over explicit naming in this enumeration.
+#[derive(
+	Copy,
+	Clone,
+	Eq,
+	PartialEq,
+	Ord,
+	PartialOrd,
+	Encode,
+	Decode,
+	Debug,
+	TypeInfo,
+	MaxEncodedLen,
+	Serialize,
+	Deserialize,
+)]
+pub enum NetworkId {
+	/// Network specified by the first 32 bytes of its genesis block.
+	ByGenesis([u8; 32]),
+	/// Network defined by the first 32-bytes of the hash and number of some block it contains.
+	ByFork { block_number: u64, block_hash: [u8; 32] },
+	/// The Polkadot mainnet Relay-chain.
+	Polkadot,
+	/// The Kusama canary-net Relay-chain.
+	Kusama,
+	/// An Ethereum network specified by its chain ID.
+	Ethereum {
+		/// The EIP-155 chain ID.
+		#[codec(compact)]
+		chain_id: u64,
+	},
+	/// The Bitcoin network, including hard-forks supported by Bitcoin Core development team.
+	BitcoinCore,
+	/// The Bitcoin network, including hard-forks supported by Bitcoin Cash developers.
+	BitcoinCash,
+	/// The Polkadot Bulletin chain.
+	PolkadotBulletin,
+}
+
+impl From<OldNetworkId> for Option<NetworkId> {
+	fn from(old: OldNetworkId) -> Self {
+		Some(NetworkId::from(old))
+	}
+}
+
+impl From<OldNetworkId> for NetworkId {
+	fn from(old: OldNetworkId) -> Self {
+		use OldNetworkId::*;
+		match old {
+			ByGenesis(hash) => Self::ByGenesis(hash),
+			ByFork { block_number, block_hash } => Self::ByFork { block_number, block_hash },
+			Polkadot => Self::Polkadot,
+			Kusama => Self::Kusama,
+			Westend => Self::ByGenesis(WESTEND_GENESIS_HASH),
+			Rococo => Self::ByGenesis(ROCOCO_GENESIS_HASH),
+			Wococo => Self::ByGenesis(WOCOCO_GENESIS_HASH),
```

Comment:
```suggestion
			Wococo => unimplemented!(),
```

Reply:
This will panic and it will panic in the runtime if you receive a message like this. I would instead either turn this into `TryFrom` or return a nonsensical by genesis network id

Reply:
I was thinking nobody uses Wococo, but you are right that someone could intentionally send a malicious V4 XCM with an internal `NetworkId::Wococo` so that conversion fails and panics the runtime.
Will revert! 👍

Reply:
oh, right, good point